js/modules/messages_data_migrator.js

```diff
@@ -61,20 +74,154 @@ const processNext = async ({
+  getNumMessages({ connection }).then((numMessages) => {
+    numTotalMessages = numMessages;
+  });
+
+  const migrationStartTime = Date.now();
+  let unprocessedMessages = [];
+  let totalMessagesProcessed = 0;
+  do {
+    const lastProcessedIndex =
+      // eslint-disable-next-line no-await-in-loop
+      await settings.getAttachmentMigrationLastProcessedIndex(connection);
+
+    const fetchUnprocessedMessagesStartTime = Date.now();
+    unprocessedMessages =
+      // eslint-disable-next-line no-await-in-loop
+      await _dangerouslyFetchMessagesRequiringSchemaUpgradeWithoutIndex({
+        connection,
+        count: NUM_MESSAGES_PER_BATCH,
+        lastIndex: lastProcessedIndex,
+      });
+    const fetchDuration = Date.now() - fetchUnprocessedMessagesStartTime;
+    const numUnprocessedMessages = unprocessedMessages.length;
+
+    if (numUnprocessedMessages === 0) {
+      break;
+    }
+
+    const upgradeStartTime = Date.now();
+    const upgradedMessages =
+      // eslint-disable-next-line no-await-in-loop
+      await Promise.all(unprocessedMessages.map(upgradeMessageSchema));
+    const upgradeDuration = Date.now() - upgradeStartTime;
+
+    const saveMessagesStartTime = Date.now();
+    const transaction = connection.transaction(MESSAGES_STORE_NAME, 'readwrite');
+    const transactionCompletion = database.completeTransaction(transaction);
+    // eslint-disable-next-line no-await-in-loop
+    await Promise.all(upgradedMessages.map(_saveMessage({ transaction })));
+    // eslint-disable-next-line no-await-in-loop
+    await transactionCompletion;
+    const saveDuration = Date.now() - saveMessagesStartTime;
+
+    // TODO: Confirm transaction is complete
+
+    const lastMessage = last(upgradedMessages);
+    const newLastProcessedIndex = lastMessage ? lastMessage.id : null;
+    if (newLastProcessedIndex) {
+      // eslint-disable-next-line no-await-in-loop
+      await settings.setAttachmentMigrationLastProcessedIndex(
+        connection,
+        newLastProcessedIndex
+      );
+    }
+
+    totalMessagesProcessed += numUnprocessedMessages;
+    console.log('Upgrade message schema:', {
+      lastProcessedIndex,
+      numUnprocessedMessages,
+      numCumulativeMessagesProcessed: totalMessagesProcessed,
+      numTotalMessages,
+      fetchDuration,
+      saveDuration,
+      upgradeDuration,
+      newLastProcessedIndex,
+      targetSchemaVersion: Message.CURRENT_SCHEMA_VERSION,
+    });
+  } while (unprocessedMessages.length > 0);
+
+  await settings.markAttachmentMigrationComplete(connection);
+  await settings.deleteAttachmentMigrationLastProcessedIndex(connection);
+
+  console.log('Close database connection');
+  connection.close();
+
+  const totalDuration = Date.now() - migrationStartTime;
+  console.log('Attachment migration complete:', {
+    totalDuration,
+    totalMessagesProcessed,
+  });
+};
+
+const _saveMessageBackbone = ({ BackboneMessage } = {}) => (message) => {
```

Comment:
What benefit to we get from extracting this out from its single callsite? I was upfile, in `processNext()` and I needed to come down here and see what this function was doing.